File: libs/east-py/packages/east-py-io/east_py_io/storage/s3.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""S3 platform functions for East.

Provides S3 and S3-compatible object storage operations for East programs,
including upload, download, delete, list, and presigned URL generation.  The
``s3_*_impl`` functions are plain Python callables taking and returning East
values and can be called directly from project code without an IR round-trip.
"""

import importlib.util
from datetime import datetime
from typing import Any

from east.runtime.platform import platform_function, platform_functions

_HAS_S3_SUPPORT = importlib.util.find_spec("boto3") is not None


def _check_s3_support() -> None:
    """Raise if the s3 extra (boto3) isn't installed."""
    if not _HAS_S3_SUPPORT:
        raise NotImplementedError(
            "S3 support requires the 's3' extra. "
            "Add east-py-io[s3] to your pyproject.toml dependencies."
        )
from east.types.types import (
    ArrayType,
    BlobType,
    BooleanType,
    DateTimeType,
    IntegerType,
    NullType,
    OptionType,
    StringType,
    StructType,
)
from east.types.values import EastArray, EastBlob, EastStruct, EastVariant

S3ConfigType = StructType(
    [
        ("region", StringType),
        ("bucket", StringType),
        ("accessKeyId", OptionType(StringType)),
        ("secretAccessKey", OptionType(StringType)),
        ("endpoint", OptionType(StringType)),
    ]
)
"""S3 or S3-compatible object storage connection configuration.

Fields: ``region`` (``String`` - AWS region name, e.g. ``"us-east-1"``),
``bucket`` (``String``), ``accessKeyId`` (``Option<String>`` - AWS access key;
falls back to the SDK credential chain when absent),
``secretAccessKey`` (``Option<String>`` - AWS secret key; paired with
``accessKeyId``), ``endpoint`` (``Option<String>`` - custom endpoint URL for
S3-compatible stores such as MinIO or Cloudflare R2; omit for AWS).
"""

S3ObjectMetadataType = StructType(
    [
        ("key", StringType),
        ("size", IntegerType),
        ("lastModified", DateTimeType),
        ("contentType", OptionType(StringType)),
        ("etag", OptionType(StringType)),
    ]
)
"""Metadata for a single S3 object returned by ``s3_head_object`` / ``s3_list_objects``.

Fields: ``key`` (``String`` - object key), ``size`` (``Integer`` - bytes),
``lastModified`` (``DateTime``), ``contentType`` (``Option<String>`` - MIME
type; absent in list results), ``etag`` (``Option<String>`` - object ETag).
"""

S3ListResultType = StructType(
    [
        ("objects", ArrayType(S3ObjectMetadataType)),
        ("isTruncated", BooleanType),
        ("continuationToken", OptionType(StringType)),
    ]
)
"""Paginated listing result from ``s3_list_objects``.

Fields: ``objects`` (``Array<S3ObjectMetadataType>``),
``isTruncated`` (``Boolean`` - ``True`` when more pages follow),
``continuationToken`` (``Option<String>`` - pass as the
``continuation_token`` argument of the next call to retrieve the next page).
"""


def create_s3_client(config: EastStruct) -> Any:
    """Create a boto3 S3 client from an ``S3ConfigType`` struct.

    Args:
        config: ``S3ConfigType`` (``EastStruct``) - see type definition above.

    Returns:
        Configured boto3 S3 client (``botocore.client.S3``).

    Raises:
        NotImplementedError: the ``s3`` extra (boto3) is not installed.
    """
    _check_s3_support()
    import boto3
    from botocore.client import Config

    access_key_id = None
    secret_access_key = None

    access_key_opt = config["accessKeyId"]
    secret_key_opt = config["secretAccessKey"]

    if access_key_opt.type == "some":
        access_key_id = access_key_opt.value
    if secret_key_opt.type == "some":
        secret_access_key = secret_key_opt.value

    endpoint_url = None
    endpoint_opt = config["endpoint"]
    if endpoint_opt.type == "some":
        endpoint_url = endpoint_opt.value

    session = boto3.session.Session()
    client_config = Config(signature_version="s3v4")

    client_kwargs: dict[str, Any] = {
        "service_name": "s3",
        "region_name": config["region"],
        "config": client_config,
    }

    if access_key_id and secret_access_key:
        client_kwargs["aws_access_key_id"] = access_key_id
        client_kwargs["aws_secret_access_key"] = secret_access_key

    if endpoint_url:
        client_kwargs["endpoint_url"] = endpoint_url

    return session.client(**client_kwargs)


@platform_function(
    name="s3_put_object",
    inputs=[S3ConfigType, StringType, BlobType],
    output=NullType,
)
async def s3_put_object_impl(config: EastStruct, key: str, data: EastBlob) -> None:
    """Upload an object to an S3 bucket.

    Args:
        config: ``S3ConfigType`` (``EastStruct``) with fields:

            - ``region`` (``String``): AWS region.
            - ``bucket`` (``String``): destination bucket name.
            - ``accessKeyId`` (``Option<String>``): AWS access key ID.
            - ``secretAccessKey`` (``Option<String>``): AWS secret access key.
            - ``endpoint`` (``Option<String>``): custom endpoint URL for
              S3-compatible stores; omit for AWS.

        key: ``String`` - destination object key (path) within the bucket.
        data: ``Blob`` (``EastBlob``) - binary content to upload.

    Returns:
        ``Null`` on success.

    Raises:
        NotImplementedError: the ``s3`` extra (boto3) is not installed.
        Exception: S3 ``ClientError`` on upload failure.
    """
    from botocore.exceptions import ClientError

    try:
        client = create_s3_client(config)
        bucket = config["bucket"]
        client.put_object(Bucket=bucket, Key=key, Body=bytes(data))
    except ClientError as e:
        raise Exception(f"S3 putObject failed: {e}") from e


@platform_function(
    name="s3_get_object",
    inputs=[S3ConfigType, StringType],
    output=BlobType,
)
async def s3_get_object_impl(config: EastStruct, key: str) -> EastBlob:
    """Download an object from an S3 bucket.

    Args:
        config: ``S3ConfigType`` (``EastStruct``) - see :func:`s3_put_object_impl`.
        key: ``String`` - object key (path) within the bucket.

    Returns:
        ``Blob`` (``EastBlob``) - raw object content.

    Raises:
        NotImplementedError: the ``s3`` extra (boto3) is not installed.
        Exception: S3 ``ClientError`` on download failure (e.g. key not found).
    """
    from botocore.exceptions import ClientError

    try:
        client = create_s3_client(config)
        bucket = config["bucket"]
        response = client.get_object(Bucket=bucket, Key=key)
        return EastBlob(response["Body"].read())
    except ClientError as e:
        raise Exception(f"S3 getObject failed: {e}") from e


@platform_function(
    name="s3_head_object",
    inputs=[S3ConfigType, StringType],
    output=S3ObjectMetadataType,
)
async def s3_head_object_impl(config: EastStruct, key: str) -> EastStruct:
    """Fetch metadata for an S3 object without downloading its body.

    Args:
        config: ``S3ConfigType`` (``EastStruct``) - see :func:`s3_put_object_impl`.
        key: ``String`` - object key (path) within the bucket.

    Returns:
        ``S3ObjectMetadataType`` (``EastStruct``): ``key`` (``String``),
        ``size`` (``Integer`` bytes), ``lastModified`` (``DateTime``),
        ``contentType`` (``Option<String>``), ``etag`` (``Option<String>``).

    Raises:
        NotImplementedError: the ``s3`` extra (boto3) is not installed.
        Exception: S3 ``ClientError`` on failure (e.g. key not found).
    """
    from botocore.exceptions import ClientError

    try:
        client = create_s3_client(config)
        bucket = config["bucket"]
        response = client.head_object(Bucket=bucket, Key=key)

        content_type: EastVariant = (
            EastVariant("some", response["ContentType"])
            if response.get("ContentType")
            else EastVariant("none", None)
        )
        etag: EastVariant = (
            EastVariant("some", response["ETag"])
            if response.get("ETag")
            else EastVariant("none", None)
        )

        return EastStruct(
            {
                "key": key,
                "size": response.get("ContentLength", 0),
                "lastModified": response.get("LastModified", datetime.now()),
                "contentType": content_type,
                "etag": etag,
            }
        )
    except ClientError as e:
        raise Exception(f"S3 headObject failed: {e}") from e


@platform_function(
    name="s3_delete_object",
    inputs=[S3ConfigType, StringType],
    output=NullType,
)
async def s3_delete_object_impl(config: EastStruct, key: str) -> None:
    """Delete an object from an S3 bucket.

    Args:
        config: ``S3ConfigType`` (``EastStruct``) - see :func:`s3_put_object_impl`.
        key: ``String`` - object key (path) to delete.

    Returns:
        ``Null`` on success.

    Raises:
        NotImplementedError: the ``s3`` extra (boto3) is not installed.
        Exception: S3 ``ClientError`` on deletion failure.
    """
    from botocore.exceptions import ClientError

    try:
        client = create_s3_client(config)
        bucket = config["bucket"]
        client.delete_object(Bucket=bucket, Key=key)
    except ClientError as e:
        raise Exception(f"S3 deleteObject failed: {e}") from e


@platform_function(
    name="s3_list_objects",
    inputs=[S3ConfigType, StringType, IntegerType, OptionType(StringType)],
    output=S3ListResultType,
)
async def s3_list_objects_impl(
    config: EastStruct, prefix: str, max_keys: int, continuation_token: EastVariant
) -> EastStruct:
    """List objects in an S3 bucket filtered by a key prefix.

    Supports pagination via ``continuation_token``; ``max_keys`` is clamped
    to the S3 range of 1-1000.

    Args:
        config: ``S3ConfigType`` (``EastStruct``) - see :func:`s3_put_object_impl`.
        prefix: ``String`` - key prefix to filter results.
        max_keys: ``Integer`` - maximum objects to return per page; clamped
            to ``[1, 1000]``.
        continuation_token: ``Option<String>`` (``EastVariant``) - token
            from a previous call's ``continuationToken`` field; pass
            ``none`` for the first page.

    Returns:
        ``S3ListResultType`` (``EastStruct``): ``objects``
        (``Array<S3ObjectMetadataType>``), ``isTruncated`` (``Boolean``),
        ``continuationToken`` (``Option<String>`` - present when
        ``isTruncated`` is ``True``).

    Raises:
        NotImplementedError: the ``s3`` extra (boto3) is not installed.
        Exception: S3 ``ClientError`` on listing failure.
    """
    from botocore.exceptions import ClientError

    try:
        client = create_s3_client(config)
        bucket = config["bucket"]

        clamped_max_keys = max(1, min(1000, max_keys))

        kwargs: dict[str, Any] = {
            "Bucket": bucket,
            "Prefix": prefix,
            "MaxKeys": clamped_max_keys,
        }
        if continuation_token.type == "some":
            kwargs["ContinuationToken"] = continuation_token.value

        response = client.list_objects_v2(**kwargs)

        objects: EastArray = EastArray(S3ObjectMetadataType, [])
        for obj in response.get("Contents", []):
            etag: EastVariant = (
                EastVariant("some", obj["ETag"]) if obj.get("ETag") else EastVariant("none", None)
            )

            objects.push_last(
                EastStruct(
                    {
                        "key": obj.get("Key", ""),
                        "size": obj.get("Size", 0),
                        "lastModified": obj.get("LastModified", datetime.now()),
                        "contentType": EastVariant("none", None),
                        "etag": etag,
                    }
                )
            )

        next_token: EastVariant = (
            EastVariant("some", response["NextContinuationToken"])
            if response.get("NextContinuationToken")
            else EastVariant("none", None)
        )

        return EastStruct(
            {
                "objects": objects,
                "isTruncated": response.get("IsTruncated", False),
                "continuationToken": next_token,
            }
        )
    except ClientError as e:
        raise Exception(f"S3 listObjects failed: {e}") from e


@platform_function(
    name="s3_presign_url",
    inputs=[S3ConfigType, StringType, IntegerType],
    output=StringType,
)
async def s3_presign_url_impl(config: EastStruct, key: str, expires_in: int) -> str:
    """Generate a presigned GET URL for temporary, unauthenticated object access.

    ``expires_in`` is clamped to the S3-supported range of 1 second to 7 days
    (604800 seconds).

    Args:
        config: ``S3ConfigType`` (``EastStruct``) - see :func:`s3_put_object_impl`.
        key: ``String`` - object key (path) within the bucket.
        expires_in: ``Integer`` - URL validity period in seconds; clamped to
            ``[1, 604800]``.

    Returns:
        ``String`` - presigned URL valid for the specified duration.

    Raises:
        NotImplementedError: the ``s3`` extra (boto3) is not installed.
        Exception: S3 ``ClientError`` on URL generation failure.
    """
    from botocore.exceptions import ClientError

    try:
        client = create_s3_client(config)
        bucket = config["bucket"]

        clamped_expires_in = max(1, min(604800, expires_in))

        url = client.generate_presigned_url(
            "get_object",
            Params={"Bucket": bucket, "Key": key},
            ExpiresIn=clamped_expires_in,
        )

        return url
    except ClientError as e:
        raise Exception(f"S3 presignUrl failed: {e}") from e


# Collected from the @platform_function decorations above.
s3_impl = platform_functions(__name__)


__all__ = [
    "s3_impl",
    "s3_put_object_impl",
    "s3_get_object_impl",
    "s3_head_object_impl",
    "s3_delete_object_impl",
    "s3_list_objects_impl",
    "s3_presign_url_impl",
    "S3ConfigType",
    "S3ObjectMetadataType",
    "S3ListResultType",
]

File: libs/east-node/packages/east-node-io/src/storage/s3.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */

/**
 * S3 platform functions for East Node IO.
 *
 * Provides S3 and S3-compatible object storage operations for East programs,
 * including upload, download, delete, list, and presigned URL generation.
 *
 * @packageDocumentation
 */

import { East, BlobType, StringType, IntegerType, NullType, OptionType, variant } from "@elaraai/east";
import type { ValueTypeOf } from "@elaraai/east";
import type { PlatformFunction } from "@elaraai/east/internal";
import { EastError } from "@elaraai/east/internal";
import {
    S3Client,
    PutObjectCommand,
    GetObjectCommand,
    DeleteObjectCommand,
    ListObjectsV2Command,
    HeadObjectCommand,
    type ListObjectsV2CommandInput,
    type _Object,
} from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
import { S3ConfigType, S3ListResultType, S3ObjectMetadataType } from "./types.js";

/**
 * Creates an S3 client from configuration.
 *
 * @param config - S3 configuration
 * @returns Configured S3Client instance
 * @internal
 */
function createS3Client(config: ValueTypeOf<typeof S3ConfigType>): S3Client {
    const credentials =
        config.accessKeyId?.type === 'some' && config.secretAccessKey?.type === 'some'
            ? {
                accessKeyId: config.accessKeyId.value,
                secretAccessKey: config.secretAccessKey.value,
            }
            : undefined;

    const endpoint = config.endpoint?.type === 'some' ? config.endpoint.value : undefined;

    // Build S3 client config conditionally to satisfy exactOptionalPropertyTypes
    const clientConfig: any = {
        region: config.region,
        credentials,
    };

    if (endpoint !== undefined) {
        clientConfig.endpoint = endpoint;
        clientConfig.forcePathStyle = true; // Use path-style for S3-compatible services
    }

    return new S3Client(clientConfig);
}

/**
 * Uploads an object to S3.
 *
 * Uploads binary data to an S3 bucket with the specified key (path).
 * Overwrites existing objects with the same key.
 *
 * This is a platform function for the East language, enabling S3 object storage
 * operations in East programs running on Node.js.
 *
 * @param config - S3 configuration
 * @param key - Object key (path) in the bucket
 * @param data - Binary data to upload
 * @returns Null on success
 *
 * @throws {EastError} When upload fails due to:
 * - Invalid bucket name (location: "s3_put_object")
 * - Authentication failure (location: "s3_put_object")
 * - Network errors (location: "s3_put_object")
 * - Permission denied (location: "s3_put_object")
 *
 * @example
 * ```ts
 * import { East, StringType, BlobType } from "@elaraai/east";
 * import { Storage } from "@elaraai/east-node-io";
 *
 * const uploadFile = East.function([StringType, BlobType], StringType, ($, filename, data) => {
 *     const config = $.let({
 *         region: "us-east-1",
 *         bucket: "my-bucket",
 *         accessKeyId: East.some("AKIAIOSFODNN7EXAMPLE"),
 *         secretAccessKey: East.some("wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"),
 *         endpoint: East.none(),
 *     });
 *
 *     // Upload the file
 *     $(Storage.S3.putObject(config, filename, data));
 *
 *     // Generate presigned URL for sharing
 *     return Storage.S3.presignUrl(config, filename, 3600n);
 * });
 *
 * const compiled = East.compileAsync(uploadFile.toIR(), Storage.S3.Implementation);
 * const url = await compiled("report.pdf", pdfBlob);
 * ```
 *
 * @remarks
 * - Supports S3 and S3-compatible services (MinIO, DigitalOcean Spaces, etc.)
 * - Uses AWS SDK v3 for S3 operations
 * - All operations are asynchronous (use East.compileAsync)
 */
export const s3_put_object = East.asyncPlatform("s3_put_object", [S3ConfigType, StringType, BlobType], NullType);

/**
 * Downloads an object from S3.
 *
 * Retrieves binary data from an S3 bucket by key (path).
 * Returns the object data as a Blob (Uint8Array).
 *
 * This is a platform function for the East language, enabling S3 object storage
 * operations in East programs running on Node.js.
 *
 * @param config - S3 configuration
 * @param key - Object key (path) in the bucket
 * @returns Binary data as Blob
 *
 * @throws {EastError} When download fails due to:
 * - Object not found (location: "s3_get_object")
 * - Invalid bucket name (location: "s3_get_object")
 * - Authentication failure (location: "s3_get_object")
 * - Network errors (location: "s3_get_object")
 * - Permission denied (location: "s3_get_object")
 *
 * @example
 * ```ts
 * import { East, StringType, BlobType } from "@elaraai/east";
 * import { Storage } from "@elaraai/east-node-io";
 *
 * const downloadFile = East.function([StringType], BlobType, ($, filename) => {
 *     const config = $.let({
 *         region: "us-east-1",
 *         bucket: "my-bucket",
 *         accessKeyId: East.none(),
 *         secretAccessKey: East.none(),
 *         endpoint: East.none(),
 *     });
 *
 *     return Storage.S3.getObject(config, filename);
 * });
 *
 * const compiled = East.compileAsync(downloadFile.toIR(), Storage.S3.Implementation);
 * const pdfData = await compiled("report.pdf");  // Returns Uint8Array
 * ```
 *
 * @remarks
 * - Returns Uint8Array (BlobType) containing raw binary data
 * - Use decodeUtf8() to convert text files to strings
 * - Streams large objects automatically
 */
export const s3_get_object = East.asyncPlatform("s3_get_object", [S3ConfigType, StringType], BlobType);
/**
 * Retrieves object metadata from S3 without downloading the file.
 *
 * Gets metadata for an S3 object including size, ETag (hash), last modified time,
 * and content type. This is useful for checking if a file has changed without
 * downloading it by comparing ETags.
 *
 * This is a platform function for the East language, enabling S3 object storage
 * operations in East programs running on Node.js.
 *
 * @param config - S3 configuration
 * @param key - Object key (path) in the bucket
 * @returns Object metadata including ETag, size, and timestamps
 *
 * @throws {EastError} When metadata retrieval fails due to:
 * - Object not found (location: "s3_head_object")
 * - Invalid bucket name (location: "s3_head_object")
 * - Authentication failure (location: "s3_head_object")
 * - Network errors (location: "s3_head_object")
 * - Permission denied (location: "s3_head_object")
 *
 * @example
 * ```ts
 * import { East, StringType } from "@elaraai/east";
 * import { Storage } from "@elaraai/east-node-io";
 *
 * const checkFileChanged = East.function([StringType, StringType], BooleanType, ($, filename, localEtag) => {
 *     const config = $.let({
 *         region: "us-east-1",
 *         bucket: "my-bucket",
 *         accessKeyId: East.none(),
 *         secretAccessKey: East.none(),
 *         endpoint: East.none(),
 *     });
 *
 *     const metadata = $.let(Storage.S3.headObject(config, filename));
 *
 *     return $.match(metadata.etag, {
 *         some: ($, remoteEtag) => remoteEtag.notEqual(localEtag),
 *         none: ($) => East.value(true), // File changed if no ETag available
 *     });
 * });
 *
 * const compiled = East.compileAsync(checkFileChanged.toIR(), Storage.S3.Implementation);
 * const changed = await compiled("report.pdf", '"abc123def456"');  // ETags include quotes
 * if (changed) {
 *     // Download the file since it changed
 * }
 * ```
 *
 * @remarks
 * - ETag is typically an MD5 hash (for simple uploads) or composite hash (for multipart uploads)
 * - ETags include surrounding quotes (e.g., '"abc123"')
 * - Much faster than downloading the entire file
 * - Use this to implement efficient sync/caching logic
 */
export const s3_head_object = East.asyncPlatform("s3_head_object", [S3ConfigType, StringType], S3ObjectMetadataType);

/**
 * Deletes an object from S3.
 *
 * Removes an object from an S3 bucket by key (path).
 * Succeeds even if the object doesn't exist (idempotent).
 *
 * This is a platform function for the East language, enabling S3 object storage
 * operations in East programs running on Node.js.
 *
 * @param config - S3 configuration
 * @param key - Object key (path) to delete
 * @returns Null on success
 *
 * @throws {EastError} When deletion fails due to:
 * - Invalid bucket name (location: "s3_delete_object")
 * - Authentication failure (location: "s3_delete_object")
 * - Network errors (location: "s3_delete_object")
 * - Permission denied (location: "s3_delete_object")
 *
 * @example
 * ```ts
 * import { East, StringType, NullType } from "@elaraai/east";
 * import { Storage } from "@elaraai/east-node-io";
 *
 * const deleteFile = East.function([StringType], NullType, ($, filename) => {
 *     const config = $.let({
 *         region: "us-east-1",
 *         bucket: "my-bucket",
 *         accessKeyId: East.none(),
 *         secretAccessKey: East.none(),
 *         endpoint: East.none(),
 *     });
 *
 *     return Storage.S3.deleteObject(config, filename);
 * });
 *
 * const compiled = East.compileAsync(deleteFile.toIR(), Storage.S3.Implementation);
 * await compiled("old-report.pdf");
 * ```
 *
 * @remarks
 * - Idempotent: succeeds even if object doesn't exist
 * - Does not delete versioned objects (only current version)
 */
export const s3_delete_object = East.asyncPlatform("s3_delete_object", [S3ConfigType, StringType], NullType);

/**
 * Lists objects in an S3 bucket with a prefix.
 *
 * Retrieves metadata for objects matching a prefix, with pagination support.
 * Returns up to `maxKeys` objects per request. Pass a continuation token
 * from a previous result to fetch the next page.
 *
 * This is a platform function for the East language, enabling S3 object storage
 * operations in East programs running on Node.js.
 *
 * @param config - S3 configuration
 * @param prefix - Prefix to filter objects (empty string for all objects)
 * @param maxKeys - Maximum number of objects to return (1-1000)
 * @param continuationToken - Continuation token from a previous list result for pagination (None for first page)
 * @returns List result with objects and pagination info
 *
 * @throws {EastError} When listing fails due to:
 * - Invalid bucket name (location: "s3_list_objects")
 * - Authentication failure (location: "s3_list_objects")
 * - Network errors (location: "s3_list_objects")
 * - Permission denied (location: "s3_list_objects")
 * - Invalid maxKeys value (location: "s3_list_objects")
 *
 * @example
 * ```ts
 * import { East, StringType, ArrayType } from "@elaraai/east";
 * import { Storage, S3ObjectMetadataType } from "@elaraai/east-node-io";
 *
 * const listReports = East.function([], ArrayType(S3ObjectMetadataType), ($) => {
 *     const config = $.let({
 *         region: "us-east-1",
 *         bucket: "my-bucket",
 *         accessKeyId: East.none(),
 *         secretAccessKey: East.none(),
 *         endpoint: East.none(),
 *     });
 *
 *     // First page
 *     const result = $.let(Storage.S3.listObjects(config, "reports/", 100n, East.none()));
 *
 *     // Next page using continuation token
 *     const page2 = $.let(Storage.S3.listObjects(config, "reports/", 100n, result.continuationToken));
 *     return page2.objects;
 * });
 *
 * const compiled = East.compileAsync(listReports.toIR(), Storage.S3.Implementation);
 * const reports = await compiled();
 * ```
 *
 * @remarks
 * - Returns objects sorted by key (lexicographically)
 * - Pass the `continuationToken` from the result to fetch the next page
 * - When `isTruncated` is false, there are no more pages
 * - maxKeys is clamped to 1-1000 range
 */
export const s3_list_objects = East.asyncPlatform("s3_list_objects", [S3ConfigType, StringType, IntegerType, OptionType(StringType)], S3ListResultType);

/**
 * Generates a presigned URL for temporary access to an S3 object.
 *
 * Creates a signed URL that grants temporary access to an object without
 * requiring AWS credentials. Useful for sharing files or client-side uploads.
 *
 * This is a platform function for the East language, enabling S3 object storage
 * operations in East programs running on Node.js.
 *
 * @param config - S3 configuration
 * @param key - Object key (path) in the bucket
 * @param expiresIn - URL expiration time in seconds (1-604800)
 * @returns Presigned URL as string
 *
 * @throws {EastError} When URL generation fails due to:
 * - Invalid bucket name (location: "s3_presign_url")
 * - Invalid expiration time (location: "s3_presign_url")
 * - Missing credentials (location: "s3_presign_url")
 * - Network errors (location: "s3_presign_url")
 *
 * @example
 * ```ts
 * import { East, StringType } from "@elaraai/east";
 * import { Storage } from "@elaraai/east-node-io";
 *
 * const shareFile = East.function([StringType], StringType, ($, filename) => {
 *     const config = $.let({
 *         region: "us-east-1",
 *         bucket: "my-bucket",
 *         accessKeyId: East.some("AKIAIOSFODNN7EXAMPLE"),
 *         secretAccessKey: East.some("wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"),
 *         endpoint: East.none(),
 *     });
 *
 *     // Generate URL valid for 1 hour
 *     return Storage.S3.presignUrl(config, filename, 3600n);
 * });
 *
 * const compiled = East.compileAsync(shareFile.toIR(), Storage.S3.Implementation);
 * const url = await compiled("report.pdf");
 * console.log(`Share this link: ${url}`);
 * ```
 *
 * @remarks
 * - URL expires after `expiresIn` seconds
 * - Requires credentials (cannot use default AWS credential chain)
 * - Works with S3-compatible services
 * - Maximum expiration: 7 days (604800 seconds)
 */
export const s3_presign_url = East.asyncPlatform("s3_presign_url", [S3ConfigType, StringType, IntegerType], StringType);

/**
 * Node.js implementation of S3 platform functions.
 *
 * Provides the runtime implementations for S3 operations using AWS SDK v3.
 * Pass this to East.compileAsync() to enable S3 functionality.
 */
export const S3Impl: PlatformFunction[] = [
    s3_put_object.implement(async (
        config: ValueTypeOf<typeof S3ConfigType>,
        key: ValueTypeOf<typeof StringType>,
        data: ValueTypeOf<typeof BlobType>
    ): Promise<null> => {
        try {
            const client = createS3Client(config);
            const command = new PutObjectCommand({
                Bucket: config.bucket,
                Key: key,
                Body: data,
            });
            await client.send(command);
            return null;
        } catch (err: any) {
            throw new EastError(`S3 putObject failed: ${err.message}`, {
                location: [{ filename: "s3_put_object", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),

    s3_get_object.implement(async (
        config: ValueTypeOf<typeof S3ConfigType>,
        key: ValueTypeOf<typeof StringType>
    ): Promise<Uint8Array> => {
        try {
            const client = createS3Client(config);
            const command = new GetObjectCommand({
                Bucket: config.bucket,
                Key: key,
            });
            const response = await client.send(command);

            if (!response.Body) {
                throw new Error('Response body is empty');
            }

            // Convert stream to Uint8Array
            const chunks: Uint8Array[] = [];
            for await (const chunk of response.Body as any) {
                chunks.push(chunk);
            }

            // Concatenate all chunks
            const totalLength = chunks.reduce((acc, chunk) => acc + chunk.length, 0);
            const result = new Uint8Array(totalLength);
            let offset = 0;
            for (const chunk of chunks) {
                result.set(chunk, offset);
                offset += chunk.length;
            }

            return result;
        } catch (err: any) {
            throw new EastError(`S3 getObject failed: ${err.message}`, {
                location: [{ filename: "s3_get_object", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),

    s3_head_object.implement(async (
        config: ValueTypeOf<typeof S3ConfigType>,
        key: ValueTypeOf<typeof StringType>
    ): Promise<ValueTypeOf<typeof S3ObjectMetadataType>> => {
        try {
            const client = createS3Client(config);
            const command = new HeadObjectCommand({
                Bucket: config.bucket,
                Key: key,
            });
            const response = await client.send(command);

            return {
                key,
                size: BigInt(response.ContentLength || 0),
                lastModified: response.LastModified || new Date(),
                contentType: response.ContentType ? variant('some', response.ContentType) : variant('none', null),
                etag: response.ETag ? variant('some', response.ETag) : variant('none', null),
            };
        } catch (err: any) {
            throw new EastError(`S3 headObject failed: ${err.message}`, {
                location: [{ filename: "s3_head_object", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),

    s3_delete_object.implement(async (
        config: ValueTypeOf<typeof S3ConfigType>,
        key: ValueTypeOf<typeof StringType>
    ): Promise<null> => {
        try {
            const client = createS3Client(config);
            const command = new DeleteObjectCommand({
                Bucket: config.bucket,
                Key: key,
            });
            await client.send(command);
            return null;
        } catch (err: any) {
            throw new EastError(`S3 deleteObject failed: ${err.message}`, {
                location: [{ filename: "s3_delete_object", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),

    s3_list_objects.implement(async (
        config: ValueTypeOf<typeof S3ConfigType>,
        prefix: ValueTypeOf<typeof StringType>,
        maxKeys: ValueTypeOf<typeof IntegerType>,
        continuationToken: ValueTypeOf<ReturnType<typeof OptionType<typeof StringType>>>
    ): Promise<ValueTypeOf<typeof S3ListResultType>> => {
        try {
            const client = createS3Client(config);

            // Clamp maxKeys to valid range (1-1000)
            const maxKeysNum = Number(maxKeys);
            const clampedMaxKeys = Math.max(1, Math.min(1000, maxKeysNum));

            const commandInput: ListObjectsV2CommandInput = {
                Bucket: config.bucket,
                Prefix: prefix,
                MaxKeys: clampedMaxKeys,
                ContinuationToken: continuationToken.type === 'some' ? continuationToken.value : undefined,
            };

            const command = new ListObjectsV2Command(commandInput);

            const response = await client.send(command);

            // Convert S3 objects to East metadata format
            // Note: ContentType is not part of _Object in ListObjectsV2, only in HeadObject/GetObject
            const objects = (response.Contents || []).map((obj: _Object) => ({
                key: obj.Key || '',
                size: BigInt(obj.Size || 0),
                lastModified: obj.LastModified || new Date(),
                contentType: variant('none', null), // ListObjectsV2 doesn't include ContentType
                etag: obj.ETag
                    ? variant('some', obj.ETag)
                    : variant('none', null),
            }));

            return {
                objects,
                isTruncated: response.IsTruncated || false,
                continuationToken: response.NextContinuationToken
                    ? variant('some', response.NextContinuationToken)
                    : variant('none', null),
            };
        } catch (err: any) {
            throw new EastError(`S3 listObjects failed: ${err.message}`, {
                location: [{ filename: "s3_list_objects", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),

    s3_presign_url.implement(async (
        config: ValueTypeOf<typeof S3ConfigType>,
        key: ValueTypeOf<typeof StringType>,
        expiresIn: ValueTypeOf<typeof IntegerType>
    ): Promise<string> => {
        try {
            const client = createS3Client(config);

            // Clamp expiresIn to valid range (1 second to 7 days)
            const expiresInNum = Number(expiresIn);
            const clampedExpiresIn = Math.max(1, Math.min(604800, expiresInNum));

            const command = new GetObjectCommand({
                Bucket: config.bucket,
                Key: key,
            });

            const url = await getSignedUrl(client, command, {
                expiresIn: clampedExpiresIn,
            });

            return url;
        } catch (err: any) {
            throw new EastError(`S3 presignUrl failed: ${err.message}`, {
                location: [{ filename: "s3_presign_url", line: 0n, column: 0n }],
                cause: err
            });
        }
    }),
];
